pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -272,3 +300,13 @@ func (d *ServiceEntryStore) GetIstioServiceAccounts(svc *model.Service, ports []
 	// service, with service instances, and dns.
 	return nil
 }
+
+// This method compares if services have changed, that needs full push.
+func servicesChanged(os []*model.Service, ns []*model.Service) bool {
+	// Length of services have changed, needs full push.
+	if len(os) != len(ns) {
+		return true
+	}
+	// TODO : check if any finer grained check is possible, rather than comparing entire object.
+	return !reflect.DeepEqual(os, ns)
```

Comment:
you could optimize this further by creating a map of services on both and comparing two service objects with same hostname.

Reply:
Yeah. Good point. Let me do it in a follow-up